File: client/lib/models/role.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'role.freezed.dart';
part 'role.g.dart';

@freezed
class Permission with _$Permission {
  const factory Permission({
    required String resource,
    @Default(false) bool canCreate,
    @Default(false) bool canRead,
    @Default(false) bool canUpdate,
    @Default(false) bool canDelete,
    @Default([]) List<String> customActions,
  }) = _Permission;

  factory Permission.fromJson(Map<String, dynamic> json) =>
      _$PermissionFromJson(json);
}

@freezed
class RuleSet with _$RuleSet {
  const factory RuleSet({
    required String id,
    required String name,
    String? description,
    @Default([]) List<Map<String, dynamic>> conditions,
    @Default('allow') String action,
  }) = _RuleSet;

  factory RuleSet.fromJson(Map<String, dynamic> json) =>
      _$RuleSetFromJson(json);
}

@freezed
class Role with _$Role {
  const factory Role({
    required String id,
    required String name,
    String? description,
    String? companyId,
    @Default('member') String level,
    @Default([]) List<Permission> permissions,
    @Default([]) List<RuleSet> ruleSets,
    @Default(true) bool isActive,
    @Default(0) int memberCount,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) = _Role;

  factory Role.fromJson(Map<String, dynamic> json) => _$RoleFromJson(json);
}

File: server/internal/dto/role.go
package dto

import (
	"time"

	"github.com/google/uuid"
)

// Permission describes access rights for a resource.
// Shape matches Flutter's Permission.fromJson.
type Permission struct {
	Resource      string   `json:"resource"`
	CanCreate     bool     `json:"canCreate"`
	CanRead       bool     `json:"canRead"`
	CanUpdate     bool     `json:"canUpdate"`
	CanDelete     bool     `json:"canDelete"`
	CustomActions []string `json:"customActions"`
}

// RuleSet defines conditional access rules.
type RuleSet struct {
	ID          string                   `json:"id"`
	Name        string                   `json:"name"`
	Description string                   `json:"description"`
	Conditions  []map[string]interface{} `json:"conditions"`
	Action      string                   `json:"action"`
}

// CreateRoleRequest is the payload for creating a company role.
type CreateRoleRequest struct {
	Name        string       `json:"name"        binding:"required"`
	Description string       `json:"description"`
	Level       string       `json:"level"`
	Permissions []Permission `json:"permissions"`
	RuleSets    []RuleSet    `json:"ruleSets"`
	IsActive    *bool        `json:"isActive"`
}

// UpdateRoleRequest allows modifying a role's definition.
type UpdateRoleRequest struct {
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Level       string       `json:"level"`
	Permissions []Permission `json:"permissions"`
	RuleSets    []RuleSet    `json:"ruleSets"`
	IsActive    *bool        `json:"isActive"`
}

// RoleResponse is the public role representation.
// Uses camelCase to match Flutter's Role.fromJson.
type RoleResponse struct {
	ID          uuid.UUID    `json:"id"`
	CompanyID   uuid.UUID    `json:"companyId"`
	Name        string       `json:"name"`
	Description string       `json:"description"`
	Level       string       `json:"level"`
	IsActive    bool         `json:"isActive"`
	Permissions []Permission `json:"permissions"`
	RuleSets    []RuleSet    `json:"ruleSets"`
	MemberCount int64        `json:"memberCount"`
	CreatedAt   time.Time    `json:"createdAt"`
	UpdatedAt   time.Time    `json:"updatedAt"`
}
